File: src-tauri/src/services/task.rs
use crate::database::{get_entity_tags, remove_search_index, set_entity_tags, upsert_search_index, Database};
use crate::error::{AppError, AppResult};
use crate::models::task::{
    CreateTaskInput, Task, TaskPriority, TaskStatus, UpdateTaskInput,
};
use chrono::{DateTime, Utc};
use rusqlite::{params, OptionalExtension};
use uuid::Uuid;

pub struct TaskService<'a> {
    db: &'a Database,
}

impl<'a> TaskService<'a> {
    pub fn new(db: &'a Database) -> Self {
        Self { db }
    }

    pub fn create(&self, input: CreateTaskInput) -> AppResult<Task> {
        let id = Uuid::new_v4().to_string();
        let now = Utc::now();
        let priority = input.priority.unwrap_or(TaskPriority::Medium);
        let tags = input.tags.unwrap_or_default();

        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute(
                "INSERT INTO tasks (id, title, description, status, priority, due_at, created_at)
                 VALUES (?1, ?2, ?3, 'todo', ?4, ?5, ?6)",
                params![
                    id,
                    input.title,
                    input.description,
                    priority.as_str(),
                    input.due_at.map(|d| d.to_rfc3339()),
                    now.to_rfc3339(),
                ],
            )?;
            set_entity_tags(&tx, "task_tags", "task_id", &id, &tags)?;
            upsert_search_index(
                &tx,
                "task",
                &id,
                &input.title,
                &input.description.clone().unwrap_or_default(),
            )?;
            tx.commit()?;
            Ok(())
        })?;

        self.get(&id)
    }

    pub fn get(&self, id: &str) -> AppResult<Task> {
        self.db.with_conn(|conn| {
            let row = conn
                .query_row(
                    "SELECT id, title, description, status, priority, due_at, created_at, completed_at
                     FROM tasks WHERE id = ?1",
                    params![id],
                    |row| {
                        Ok((
                            row.get::<_, String>(0)?,
                            row.get::<_, String>(1)?,
                            row.get::<_, Option<String>>(2)?,
                            row.get::<_, String>(3)?,
                            row.get::<_, String>(4)?,
                            row.get::<_, Option<String>>(5)?,
                            row.get::<_, String>(6)?,
                            row.get::<_, Option<String>>(7)?,
                        ))
                    },
                )
                .optional()?
                .ok_or_else(|| AppError::NotFound(format!("task {id}")))?;

            let tags = get_entity_tags(conn, "task_tags", "task_id", id)?;
            Ok(row_to_task(row, tags))
        })
    }

    pub fn list(&self, status_filter: Option<&str>) -> AppResult<Vec<Task>> {
        self.db.with_conn(|conn| {
            let mut tasks = Vec::new();
            let sql = match status_filter {
                Some(_) => "SELECT id, title, description, status, priority, due_at, created_at, completed_at
                            FROM tasks WHERE status = ?1 ORDER BY created_at DESC",
                None => "SELECT id, title, description, status, priority, due_at, created_at, completed_at
                          FROM tasks ORDER BY created_at DESC",
            };
            let mut stmt = conn.prepare(sql)?;
            let rows = match status_filter {
                Some(s) => stmt.query_map(params![s], map_task_row)?,
                None => stmt.query_map([], map_task_row)?,
            };
            for row in rows {
                let (id, title, description, status, priority, due_at, created_at, completed_at) = row?;
                let tags = get_entity_tags(conn, "task_tags", "task_id", &id)?;
                tasks.push(Task {
                    id,
                    title,
                    description,
                    status: TaskStatus::from_str(&status),
                    priority: TaskPriority::from_str(&priority),
                    due_at: parse_opt_datetime(&due_at),
                    created_at: parse_datetime(&created_at)?,
                    completed_at: parse_opt_datetime(&completed_at),
                    tags,
                });
            }
            Ok(tasks)
        })
    }

    pub fn list_today(&self) -> AppResult<Vec<Task>> {
        let all = self.list(None)?;
        let today = Utc::now().date_naive();
        Ok(all
            .into_iter()
            .filter(|t| {
                t.status != TaskStatus::Done
                    && t.status != TaskStatus::Cancelled
                    && t.due_at
                        .map(|d| d.date_naive() <= today)
                        .unwrap_or(true)
            })
            .collect())
    }

    pub fn update(&self, id: &str, input: UpdateTaskInput) -> AppResult<Task> {
        let existing = self.get(id)?;
        let title = input.title.unwrap_or(existing.title);
        let description = input.description.or(existing.description);
        let prev_status = existing.status.clone();
        let status = input.status.unwrap_or(existing.status);
        let priority = input.priority.unwrap_or(existing.priority);
        let due_at = match input.due_at {
            Some(v) => v,
            None => existing.due_at,
        };
        let tags = input.tags.unwrap_or(existing.tags);
        let completed_at = if status == TaskStatus::Done && prev_status != TaskStatus::Done {
            Some(Utc::now())
        } else if status != TaskStatus::Done {
            None
        } else {
            existing.completed_at
        };

        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute(
                "UPDATE tasks SET title=?1, description=?2, status=?3, priority=?4,
                 due_at=?5, completed_at=?6 WHERE id=?7",
                params![
                    title,
                    description,
                    status.as_str(),
                    priority.as_str(),
                    due_at.map(|d| d.to_rfc3339()),
                    completed_at.map(|d| d.to_rfc3339()),
                    id,
                ],
            )?;
            set_entity_tags(&tx, "task_tags", "task_id", id, &tags)?;
            upsert_search_index(
                &tx,
                "task",
                id,
                &title,
                &description.clone().unwrap_or_default(),
            )?;
            tx.commit()?;
            Ok(())
        })?;

        self.get(id)
    }

    pub fn find_id_by_tag(&self, tag: &str) -> AppResult<Option<String>> {
        self.db.with_conn(|conn| {
            conn.query_row(
                "SELECT tt.task_id FROM task_tags tt
                 JOIN tags t ON t.id = tt.tag_id
                 WHERE t.name = ?1
                 LIMIT 1",
                params![tag],
                |row| row.get(0),
            )
            .optional()
            .map_err(AppError::from)
        })
    }

    pub fn complete(&self, id: &str) -> AppResult<Task> {
        self.update(
            id,
            UpdateTaskInput {
                status: Some(TaskStatus::Done),
                ..Default::default()
            },
        )
    }

    pub fn delete(&self, id: &str) -> AppResult<()> {
        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute("DELETE FROM tasks WHERE id = ?1", params![id])?;
            remove_search_index(&tx, "task", id)?;
            tx.commit()?;
            Ok(())
        })
    }

    pub fn count_today_progress(&self) -> AppResult<(i32, i32)> {
        let tasks = self.list_today()?;
        let total = tasks.len() as i32;
        let done = tasks
            .iter()
            .filter(|t| t.status == TaskStatus::Done)
            .count() as i32;
        Ok((done, total))
    }
}

impl Default for UpdateTaskInput {
    fn default() -> Self {
        Self {
            title: None,
            description: None,
            status: None,
            priority: None,
            due_at: None,
            tags: None,
        }
    }
}

fn map_task_row(row: &rusqlite::Row) -> rusqlite::Result<(String, String, Option<String>, String, String, Option<String>, String, Option<String>)> {
    Ok((
        row.get(0)?,
        row.get(1)?,
        row.get(2)?,
        row.get(3)?,
        row.get(4)?,
        row.get(5)?,
        row.get(6)?,
        row.get(7)?,
    ))
}

fn row_to_task(
    row: (String, String, Option<String>, String, String, Option<String>, String, Option<String>),
    tags: Vec<String>,
) -> Task {
    Task {
        id: row.0,
        title: row.1,
        description: row.2,
        status: TaskStatus::from_str(&row.3),
        priority: TaskPriority::from_str(&row.4),
        due_at: parse_opt_datetime(&row.5),
        created_at: parse_datetime(&row.6).unwrap_or_else(|_| Utc::now()),
        completed_at: parse_opt_datetime(&row.7),
        tags,
    }
}

fn parse_datetime(s: &str) -> AppResult<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(s)
        .map(|d| d.with_timezone(&Utc))
        .map_err(|e| AppError::Other(e.to_string()))
}

fn parse_opt_datetime(s: &Option<String>) -> Option<DateTime<Utc>> {
    s.as_ref()
        .and_then(|v| DateTime::parse_from_rfc3339(v).ok())
        .map(|d| d.with_timezone(&Utc))
}

File: server/internal/core/task.go
package core

import (
	"database/sql"
	"time"
)

// Task mirrors models/task.rs (camelCase JSON).
type Task struct {
	ID          string     `json:"id"`
	Title       string     `json:"title"`
	Description *string    `json:"description,omitempty"`
	Status      string     `json:"status"`
	Priority    string     `json:"priority"`
	DueAt       *time.Time `json:"dueAt,omitempty"`
	CreatedAt   time.Time  `json:"createdAt"`
	CompletedAt *time.Time `json:"completedAt,omitempty"`
	Tags        []string   `json:"tags"`
}

// CreateTaskInput / UpdateTaskInput carry optional fields; UpdateTaskInput.DueAt
// uses a double-optional (present-but-null clears the due date).
type CreateTaskInput struct {
	Title       string     `json:"title"`
	Description *string    `json:"description,omitempty"`
	Priority    *string    `json:"priority,omitempty"`
	DueAt       *time.Time `json:"dueAt,omitempty"`
	Tags        *[]string  `json:"tags,omitempty"`
}

func priorityFromDaysUntilDue(daysLeft int) string {
	if daysLeft <= 1 {
		return "high"
	}
	return "low"
}

type taskService struct{ db *DB }

func newTaskService(db *DB) *taskService { return &taskService{db: db} }

func (s *taskService) create(in CreateTaskInput) (Task, error) {
	id := newID()
	now := nowUTC()
	priority := "medium"
	if in.Priority != nil && *in.Priority != "" {
		priority = normalizePriority(*in.Priority)
	}
	tags := []string{}
	if in.Tags != nil {
		tags = *in.Tags
	}
	var dueAt any
	if in.DueAt != nil {
		dueAt = rfc3339(*in.DueAt)
	}
	tx, err := s.db.sql.Begin()
	if err != nil {
		return Task{}, err
	}
	defer tx.Rollback()
	if _, err := tx.Exec(
		`INSERT INTO tasks (id, title, description, status, priority, due_at, created_at)
		 VALUES (?, ?, ?, 'todo', ?, ?, ?)`,
		id, in.Title, in.Description, priority, dueAt, rfc3339(now)); err != nil {
		return Task{}, err
	}
	if err := setEntityTags(tx, "task_tags", "task_id", id, tags); err != nil {
		return Task{}, err
	}
	desc := ""
	if in.Description != nil {
		desc = *in.Description
	}
	if err := upsertSearchIndex(tx, "task", id, in.Title, desc); err != nil {
		return Task{}, err
	}
	if err := tx.Commit(); err != nil {
		return Task{}, err
	}
	return s.get(id)
}

func (s *taskService) get(id string) (Task, error) {
	var (
		t                   Task
		desc, dueAt, comp   sql.NullString
		status, priority    string
		created             string
	)
	err := s.db.sql.QueryRow(
		`SELECT id, title, description, status, priority, due_at, created_at, completed_at
		 FROM tasks WHERE id = ?`, id).
		Scan(&t.ID, &t.Title, &desc, &status, &priority, &dueAt, &created, &comp)
	if err == sql.ErrNoRows {
		return Task{}, notFound("task " + id)
	}
	if err != nil {
		return Task{}, err
	}
	t.Description = nullStr(desc)
	t.Status = status
	t.Priority = priority
	t.DueAt = nullTime(dueAt)
	t.CreatedAt = parseRFC3339(created)
	t.CompletedAt = nullTime(comp)
	tags, err := getEntityTags(s.db.sql, "task_tags", "task_id", id)
	if err != nil {
		return Task{}, err
	}
	t.Tags = tags
	return t, nil
}

func (s *taskService) list(statusFilter *string) ([]Task, error) {
	var rows *sql.Rows
	var err error
	if statusFilter != nil {
		rows, err = s.db.sql.Query(
			`SELECT id, title, description, status, priority, due_at, created_at, completed_at
			 FROM tasks WHERE status = ? ORDER BY created_at DESC`, *statusFilter)
	} else {
		rows, err = s.db.sql.Query(
			`SELECT id, title, description, status, priority, due_at, created_at, completed_at
			 FROM tasks ORDER BY created_at DESC`)
	}
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []Task{}
	var ids []string
	var staged []Task
	for rows.Next() {
		var (
			t                Task
			desc, dueAt, comp sql.NullString
			status, priority string
			created          string
		)
		if err := rows.Scan(&t.ID, &t.Title, &desc, &status, &priority, &dueAt, &created, &comp); err != nil {
			return nil, err
		}
		t.Description = nullStr(desc)
		t.Status = status
		t.Priority = priority
		t.DueAt = nullTime(dueAt)
		t.CreatedAt = parseRFC3339(created)
		t.CompletedAt = nullTime(comp)
		staged = append(staged, t)
		ids = append(ids, t.ID)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	for _, t := range staged {
		tags, err := getEntityTags(s.db.sql, "task_tags", "task_id", t.ID)
		if err != nil {
			return nil, err
		}
		t.Tags = tags
		out = append(out, t)
	}
	return out, nil
}

func (s *taskService) listToday() ([]Task, error) {
	all, err := s.list(nil)
	if err != nil {
		return nil, err
	}
	today := nowUTC().Truncate(24 * time.Hour)
	out := []Task{}
	for _, t := range all {
		if t.Status == "done" || t.Status == "cancelled" {
			continue
		}
		if t.DueAt == nil || !dateAfter(*t.DueAt, today) {
			out = append(out, t)
		}
	}
	return out, nil
}

// dateAfter reports whether a's UTC date is strictly after b's UTC date.
func dateAfter(a, b time.Time) bool {
	ad := time.Date(a.Year(), a.Month(), a.Day(), 0, 0, 0, 0, time.UTC)
	bd := time.Date(b.Year(), b.Month(), b.Day(), 0, 0, 0, 0, time.UTC)
	return ad.After(bd)
}

// UpdateTaskInput uses tri-state pointers decoded manually in rpc.go so that
// "due_at present but null" clears the date. Here nil means "leave unchanged".
type UpdateTaskInput struct {
	Title       *string
	Description *string
	descSet     bool
	Status      *string
	Priority    *string
	DueAt       *time.Time
	dueSet      bool
	Tags        *[]string
}

func (s *taskService) update(id string, in UpdateTaskInput) (Task, error) {
	existing, err := s.get(id)
	if err != nil {
		return Task{}, err
	}
	title := existing.Title
	if in.Title != nil {
		title = *in.Title
	}
	desc := existing.Description
	if in.descSet {
		desc = in.Description
	}
	prevStatus := existing.Status
	status := existing.Status
	if in.Status != nil {
		status = normalizeStatus(*in.Status)
	}
	priority := existing.Priority
	if in.Priority != nil {
		priority = normalizePriority(*in.Priority)
	}
	dueAt := existing.DueAt
	if in.dueSet {
		dueAt = in.DueAt
	}
	tags := existing.Tags
	if in.Tags != nil {
		tags = *in.Tags
	}
	var completed *time.Time
	switch {
	case status == "done" && prevStatus != "done":
		n := nowUTC()
		completed = &n
	case status != "done":
		completed = nil
	default:
		completed = existing.CompletedAt
	}
	tx, err := s.db.sql.Begin()
	if err != nil {
		return Task{}, err
	}
	defer tx.Rollback()
	var dueVal, compVal any
	if dueAt != nil {
		dueVal = rfc3339(*dueAt)
	}
	if completed != nil {
		compVal = rfc3339(*completed)
	}
	if _, err := tx.Exec(
		`UPDATE tasks SET title=?, description=?, status=?, priority=?, due_at=?, completed_at=? WHERE id=?`,
		title, desc, status, priority, dueVal, compVal, id); err != nil {
		return Task{}, err
	}
	if err := setEntityTags(tx, "task_tags", "task_id", id, tags); err != nil {
		return Task{}, err
	}
	d := ""
	if desc != nil {
		d = *desc
	}
	if err := upsertSearchIndex(tx, "task", id, title, d); err != nil {
		return Task{}, err
	}
	if err := tx.Commit(); err != nil {
		return Task{}, err
	}
	return s.get(id)
}

func (s *taskService) complete(id string) (Task, error) {
	done := "done"
	return s.update(id, UpdateTaskInput{Status: &done})
}

func (s *taskService) delete(id string) error {
	tx, err := s.db.sql.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if _, err := tx.Exec("DELETE FROM tasks WHERE id = ?", id); err != nil {
		return err
	}
	if err := removeSearchIndex(tx, "task", id); err != nil {
		return err
	}
	return tx.Commit()
}

func (s *taskService) findIDByTag(tag string) (string, bool, error) {
	var id string
	err := s.db.sql.QueryRow(
		`SELECT tt.task_id FROM task_tags tt JOIN tags t ON t.id = tt.tag_id WHERE t.name = ? LIMIT 1`, tag).
		Scan(&id)
	if err == sql.ErrNoRows {
		return "", false, nil
	}
	if err != nil {
		return "", false, err
	}
	return id, true, nil
}

func (s *taskService) countTodayProgress() (int, int, error) {
	tasks, err := s.listToday()
	if err != nil {
		return 0, 0, err
	}
	total := len(tasks)
	done := 0
	for _, t := range tasks {
		if t.Status == "done" {
			done++
		}
	}
	return done, total, nil
}

func normalizeStatus(s string) string {
	switch s {
	case "doing", "done", "cancelled":
		return s
	default:
		return "todo"
	}
}

func normalizePriority(s string) string {
	switch s {
	case "low", "high":
		return s
	default:
		return "medium"
	}
}
